filament/backend/src/vulkan/memory/ResourceCounter.h

```diff
@@ -0,0 +1,337 @@
+            size_t const oldSize = mCounters.size();
+            mCounters.resize(!oldSize ? (size_t) (mArenaSize / 2) : (size_t) oldSize * 1.5);
+            return {oldSize, mCounters.size()};
+        };
+        auto const addFree = [&](size_t from, size_t to) {
+            for (size_t i = from; i < to; ++i) {
+                mFreeList.push_back(i);
+            }
+        };
+
+        if constexpr (THREAD_SAFE) {
+            size_t countersCount = 0;
+            size_t oldSize = 0;
+            {
+                std::unique_lock<utils::Mutex> lock(mCountersMutex);
+                auto [olds, counters] = expandImpl();
+                oldSize = olds;
+                countersCount = counters;
+            }
+            {
+                std::unique_lock<utils::Mutex> lock(mFreeListMutex);
+                addFree(oldSize, countersCount);
+            }
+        } else {
+            auto [olds, counters] = expandImpl();
+            addFree(olds, counters);
+        }
+    }
+
+    size_t const mArenaSize;
+    utils::Mutex mCountersMutex;
+    std::vector<CounterImpl<THREAD_SAFE>> mCounters;
+    utils::Mutex mFreeListMutex;
+    std::vector<int32_t> mFreeList;
+};
+
+// This wraps a counter, either thread-safe or not.
+struct RefCounter {
+private:
+    using TSCounterPtr = CounterImpl<IS_THREAD_SAFE>*;
+    using CounterPtr = CounterImpl<IS_NOT_THREAD_SAFE>*;
+
+public:
+    RefCounter() = default;
+
+    RefCounter(TSCounterPtr counter)
+        : mTsCounter{counter} {}
+
+    RefCounter(CounterPtr counter)
+        : mCounter{counter} {}
+
+    RefCounter(RefCounter&& rhs) = default;
+    RefCounter& operator=(RefCounter&& rhs) = default;
+
+    RefCounter(RefCounter const& rhs) = default;
+    RefCounter& operator=(RefCounter const& rhs) = default;
+
+    inline void inc() {
+        assert_invariant(bool(*this));
+        if (mTsCounter) {
+            mTsCounter->inc();
+            return;
+        }
+        mCounter->inc();
+    }
+    inline void dec() {
+        assert_invariant(bool(*this));
+        if (mTsCounter) {
+            mTsCounter->dec();
+            return;
+        }
+        mCounter->dec();
+    }
+    inline HandleId id() const {
+        assert_invariant(bool(*this));
+        if (mTsCounter) {
+            return mTsCounter->id;
+        }
+        return mCounter->id;
+    }
+
+    inline ResourceType type() const {
+        assert_invariant(bool(*this));
+        if (mTsCounter) {
+            return mTsCounter->type;
+        }
+        return mCounter->type;
+    }
+
+    inline explicit operator bool() const {
+        return mTsCounter || mCounter;
+    }
+
+    inline bool operator==(RefCounter const& rhs) const {
+        return mTsCounter == rhs.mTsCounter && mCounter == rhs.mCounter;
+    }
+
+    inline bool operator!=(RefCounter const& rhs) const {
+        return !(*this == rhs);
+    }
+
+private:
+    TSCounterPtr mTsCounter = nullptr;
+    CounterPtr mCounter = nullptr;
```

Comment:
Why can't the RefCounter be templated? This is weird that would have both pointers? This doubles its size for not reason?

Reply:
done